internals/overlord/planstate/manager.go

```diff
@@ -0,0 +1,192 @@
+	m.planLock.Lock()
+	defer m.planLock.Unlock()
+	return m.plan, nil
+}
+
+// LayerCreator allows an Append or Combine operation on the plan to include
+// layer creation derived from the plan itself, which has to happen while
+// the plan is locked.
+type LayerCreator interface {
+	Layer(planIn *plan.Plan) (*plan.Layer, error)
+}
+
+// AppendLayer takes a LayerCreator interface, which may be an existing plan
+// Layer, and appends it to the plan's layers and updates the layer.Order
+// field to the new order. If a layer with layer.Label already exists, return
+// an error of type *LabelExists.
+func (m *PlanManager) AppendLayer(creator LayerCreator) error {
+	m.planLock.Lock()
+	defer m.planLock.Unlock()
+
+	layer, err := creator.Layer(m.plan)
+	if err != nil {
+		return err
+	}
+
+	index, _ := findLayer(m.plan.Layers, layer.Label)
+	if index >= 0 {
+		return &LabelExists{Label: layer.Label}
+	}
+
+	return m.appendLayer(layer)
+}
+
+// CombineLayer takes a LayerCreator interface, which may be an existing plan
+// Layer, and appends it to an existing layer that has the same label. If no
+// existing layer has the label, append a new one. In either case, update the
+// layer.Order field to the new order.
+func (m *PlanManager) CombineLayer(creator LayerCreator) error {
+	m.planLock.Lock()
+	defer m.planLock.Unlock()
+
+	layer, err := creator.Layer(m.plan)
+	if err != nil {
+		return err
+	}
+
+	index, found := findLayer(m.plan.Layers, layer.Label)
+	if index < 0 {
+		// No layer found with this label, append new one.
+		return m.appendLayer(layer)
+	}
+
+	// Layer found with this label, combine into that one.
+	combined, err := plan.CombineLayers(found, layer)
+	if err != nil {
+		return err
+	}
+	combined.Order = found.Order
+	combined.Label = found.Label
+
+	// Insert combined layer back into plan's layers list.
+	newLayers := make([]*plan.Layer, len(m.plan.Layers))
+	copy(newLayers, m.plan.Layers)
+	newLayers[index] = combined
+	err = m.updatePlanLayers(newLayers)
+	if err != nil {
+		return err
+	}
+	layer.Order = found.Order
+	return nil
+}
+
+func (m *PlanManager) appendLayer(layer *plan.Layer) error {
+	newOrder := 1
+	if len(m.plan.Layers) > 0 {
+		last := m.plan.Layers[len(m.plan.Layers)-1]
+		newOrder = last.Order + 1
+	}
+
+	newLayers := append(m.plan.Layers, layer)
+	err := m.updatePlanLayers(newLayers)
+	if err != nil {
+		return err
+	}
+	layer.Order = newOrder
+	return nil
+}
+
+func (m *PlanManager) updatePlanLayers(layers []*plan.Layer) error {
+	combined, err := plan.CombineLayers(layers...)
+	if err != nil {
+		return err
+	}
+	plan := &plan.Plan{
+		Layers:     layers,
+		Services:   combined.Services,
+		Checks:     combined.Checks,
+		LogTargets: combined.LogTargets,
+	}
+	m.planChanged(plan)
+	return nil
+}
+
+// findLayer returns the index (in layers) of the layer with the given label,
+// or returns -1, nil if there's no layer with that label.
+func findLayer(layers []*plan.Layer, label string) (int, *plan.Layer) {
```

Comment:
*nit*: Is that idiomatic Go (returning `-1` meaning "error") or would a return type of `(int, *plan.Layer, error)` be more Go-ish?

However, I see this is mostly just migrated from old code, so feel free to ignore.

Reply:
This is taken as is, so in general I would not change existing functional code, unless something is broken (or a real need to do so). I think its perhaps taking inspiration from https://pkg.go.dev/strings#Index. We can of course improve it if needed, but feels like a separate PR as it will affect some code and tests. Going to resolve this one for now, just to keep the current discussion points focused on only plan manager changes.